Webjars broken with Quarkus 3.16 due to an exception in Classgraph

In code that worked with Quarkus 3.15, I am now getting the following exception after upgrading to Quarkus 3.16:

    java.lang.RuntimeException: Exception while calling findClassLoaderOrder for nonapi.io.github.classgraph.classloaderhandler.QuarkusClassLoaderHandler
	at nonapi.io.github.classgraph.classloaderhandler.ClassLoaderHandlerRegistry$ClassLoaderHandlerRegistryEntry.findClasspathOrder(ClassLoaderHandlerRegistry.java:239)
	at nonapi.io.github.classgraph.classpath.ClasspathFinder.<init>(ClasspathFinder.java:277)
	at io.github.classgraph.Scanner.<init>(Scanner.java:170)
	at io.github.classgraph.ClassGraph.scanAsync(ClassGraph.java:1562)
	at io.github.classgraph.ClassGraph.scanAsync(ClassGraph.java:1590)
	at io.github.classgraph.ClassGraph.scan(ClassGraph.java:1615)
	at io.github.classgraph.ClassGraph.scan(ClassGraph.java:1654)
	at io.github.classgraph.ClassGraph.scan(ClassGraph.java:1667)
	at org.webjars.WebJarAssetLocator.scanForWebJars(WebJarAssetLocator.java:188)
	at org.webjars.WebJarAssetLocator.<init>(WebJarAssetLocator.java:210)
	at org.webjars.WebJarAssetLocator.<init>(WebJarAssetLocator.java:198)
	...
    Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:115)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at nonapi.io.github.classgraph.classloaderhandler.ClassLoaderHandlerRegistry$ClassLoaderHandlerRegistryEntry.findClasspathOrder(ClassLoaderHandlerRegistry.java:236)
	... 21 more
    Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Collection.iterator()" because the return value of "nonapi.io.github.classgraph.reflection.ReflectionUtils.getFieldVal(boolean, Object, String)" is null
	at nonapi.io.github.classgraph.classloaderhandler.QuarkusClassLoaderHandler.findClasspathOrderForQuarkusClassloader(QuarkusClassLoaderHandler.java:120)
	at nonapi.io.github.classgraph.classloaderhandler.QuarkusClassLoaderHandler.findClasspathOrder(QuarkusClassLoaderHandler.java:111)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 23 more

The following code in Classgraph seems to trigger the exception:

    for (final Object element : (Collection<Object>) classpathOrder.reflectionUtils.getFieldVal(false,
                classLoader, "elements")) {
         ...
    }

I'm assuming some behavior of the Quarkus ClassLoader changed in 3.16? I do not know any details about this though.

I attempted manually forcing the latest version of Classgraph (4.8.177), but exception persists.

    $ java -version
    openjdk version "21.0.5" 2024-10-15 LTS
    OpenJDK Runtime Environment Temurin-21.0.5+11 (build 21.0.5+11-LTS)
    OpenJDK 64-Bit Server VM Temurin-21.0.5+11 (build 21.0.5+11-LTS, mixed mode, sharing)

--- a/src/main/java/nonapi/io/github/classgraph/classloaderhandler/QuarkusClassLoaderHandler.java
+++ b/src/main/java/nonapi/io/github/classgraph/classloaderhandler/QuarkusClassLoaderHandler.java
@@ -31,6 +31,7 @@ package nonapi.io.github.classgraph.classloaderhandler;
 import java.io.IOError;
 import java.net.URI;
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Map;
 
@@ -117,8 +118,17 @@ class QuarkusClassLoaderHandler implements ClassLoaderHandler {
     @SuppressWarnings("unchecked")
     private static void findClasspathOrderForQuarkusClassloader(final ClassLoader classLoader,
             final ClasspathOrder classpathOrder, final ScanSpec scanSpec, final LogNode log) {
-        for (final Object element : (Collection<Object>) classpathOrder.reflectionUtils.getFieldVal(false,
-                classLoader, "elements")) {
+
+        Collection<Object> elements = (Collection<Object>) classpathOrder.reflectionUtils.getFieldVal(false,
+            classLoader, "elements");
+        if(elements == null) {
+            elements = new ArrayList<>();
+            for(String fieldName : new String[] {"normalPriorityElements", "lesserPriorityElements"}) {
+                elements.addAll((Collection<Object>) classpathOrder.reflectionUtils.getFieldVal(false,
+                    classLoader, fieldName));
+            }
+        }
+        for (final Object element : elements) {
             final String elementClassName = element.getClass().getName();
             if ("io.quarkus.bootstrap.classloading.JarClassPathElement".equals(elementClassName)) {
                 classpathOrder.addClasspathEntry(classpathOrder.reflectionUtils.getFieldVal(false, element, "file"),

--- a/src/main/java/nonapi/io/github/classgraph/classloaderhandler/QuarkusClassLoaderHandler.java
+++ b/src/main/java/nonapi/io/github/classgraph/classloaderhandler/QuarkusClassLoaderHandler.java
@@ -115,19 +115,10 @@ class QuarkusClassLoaderHandler implements ClassLoaderHandler {
         }
     }
 
-    @SuppressWarnings("unchecked")
     private static void findClasspathOrderForQuarkusClassloader(final ClassLoader classLoader,
             final ClasspathOrder classpathOrder, final ScanSpec scanSpec, final LogNode log) {
 
-        Collection<Object> elements = (Collection<Object>) classpathOrder.reflectionUtils.getFieldVal(false,
-            classLoader, "elements");
-        if(elements == null) {
-            elements = new ArrayList<>();
-            for(String fieldName : new String[] {"normalPriorityElements", "lesserPriorityElements"}) {
-                elements.addAll((Collection<Object>) classpathOrder.reflectionUtils.getFieldVal(false,
-                    classLoader, fieldName));
-            }
-        }
+        Collection<Object> elements = findQuarkusClassLoaderElements(classLoader, classpathOrder);
         for (final Object element : elements) {
             final String elementClassName = element.getClass().getName();
             if ("io.quarkus.bootstrap.classloading.JarClassPathElement".equals(elementClassName)) {
@@ -143,6 +134,25 @@ class QuarkusClassLoaderHandler implements ClassLoaderHandler {
                 }
             }
         }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static Collection<Object> findQuarkusClassLoaderElements(ClassLoader classLoader, ClasspathOrder classpathOrder) {
+        Collection<Object> elements = (Collection<Object>) classpathOrder.reflectionUtils.getFieldVal(false,
+            classLoader, "elements");
+        if (elements == null) {
+            elements = new ArrayList<>();
+            // Since 3.16.x
+            for (String fieldName : new String[] {"normalPriorityElements", "lesserPriorityElements"}) {
+                Collection<Object> fieldVal = (Collection<Object>) classpathOrder.reflectionUtils.getFieldVal(false,
+                    classLoader, fieldName);
+                if (fieldVal == null) {
+                    continue;
+                }
+                elements.addAll(fieldVal);
+            }
+        }
+        return elements;
     }
 
     @SuppressWarnings("unchecked")

